crates/lib3h/src/engine/real_engine.rs

```diff
@@ -555,6 +554,28 @@ impl<T: Transport, D: Dht> RealEngine<T, D> {
         Lib3hServerProtocol::SuccessResult(response)
     }
 
+    /// Destroy gateway for this agent in this space, if part of it.
+    /// Respond with FailureResult if space was not already joined.
+    /// Must not already be part of this space.
+    fn serve_LeaveSpace(&mut self, join_msg: &SpaceData) -> Lib3hServerProtocol {
+        // Prepare response
+        let mut res = GenericResultData {
+            request_id: join_msg.request_id.clone(),
+            space_address: join_msg.space_address.clone(),
+            to_agent_id: join_msg.agent_id.clone(),
+            result_info: vec![],
+        };
+        // Bail if space not joined by agent
+        let chain_id = (join_msg.space_address.clone(), join_msg.agent_id.clone());
+        if !self.space_gateway_map.contains_key(&chain_id) {
+            res.result_info = "Agent is not part of the space".to_string().into_bytes();
+            return Lib3hServerProtocol::FailureResult(res);
+        }
+        self.space_gateway_map.remove(&chain_id).unwrap();
```

Comment:
Minor optimization but you could just call `remove` and check if the return value is `None`- if it is then produce the error result. Also, this code is simple enough that using a mutable reference for the result instance isn't a huge deal but it would be really easy to make this code immutable by building the result at the very end and using a conditional let statement to populate the `result_info` .

Reply:
@struktured Yeah, its a bit cleaner like you say even if the method is still not immutable.